clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -314,68 +369,92 @@ public void poll(long timeout) throws IOException {
+        }
+        try {
+            for (SelectionKey key : inHandlingOrder) {
+                keysHandled.add(key);
+                KafkaChannel channel = channel(key);
+
+                // register all per-connection metrics at once
+                sensors.maybeRegisterConnectionMetrics(channel.id());
+                if (idleExpiryManager != null)
+                    idleExpiryManager.update(channel.id(), currentTimeNanos);
+
+                try {
+
+                    /* complete any connections that have finished their handshake (either normally or immediately) */
+                    if (isImmediatelyConnected || key.isConnectable()) {
+                        if (channel.finishConnect()) {
+                            this.connected.add(channel.id());
+                            this.sensors.connectionCreated.record();
+                            SocketChannel socketChannel = (SocketChannel) key.channel();
+                            log.debug("Created socket with SO_RCVBUF = {}, SO_SNDBUF = {}, SO_TIMEOUT = {} to node {}",
+                                    socketChannel.socket().getReceiveBufferSize(),
+                                    socketChannel.socket().getSendBufferSize(),
+                                    socketChannel.socket().getSoTimeout(),
+                                    channel.id());
+                        } else
+                            continue;
+                    }
 
-                /* if channel is ready read from any connections that have readable data */
-                if (channel.ready() && key.isReadable() && !hasStagedReceive(channel)) {
-                    NetworkReceive networkReceive;
-                    while ((networkReceive = channel.read()) != null)
-                        addToStagedReceives(channel, networkReceive);
-                }
+                    /* if channel is not ready finish prepare */
+                    if (channel.isConnected() && !channel.ready())
+                        channel.prepare();
+
+                    /* if channel is ready read from any connections that have readable data */
+                    if (channel.ready() && key.isReadable() && !hasStagedReceive(channel)) {
+                        NetworkReceive networkReceive;
+                        while ((networkReceive = channel.read()) != null)
+                            addToStagedReceives(channel, networkReceive);
+                        if (channel.hasBytesBuffered()) {
+                            //this channels has bytes enqueued in intermediary buffers that we could not read
+                            //(possibly because no memory). it may be the case that the underlying socket will
+                            //not come up in the next poll() and so we need to remember this channel for the
+                            //next poll call otherwise data may be stuff in said buffers forever.
+                            keysWithBytesBuffered.add(key);
+                        }
```

Comment:
@rajinisivaram:
1. is it guaranteed that if there's anything in net buffer after a read there MUST ALWAYS be more incoming? because if so i can just react solely to data in app buffer.

2. i think i now understand the scenario you describe. my "best" idea of how to solve it would be have a boolean return value from pollSelectionKeys() to indicate if any "progress" has been made. if no progress has been made in the previous call to poll() the next call would not set timeout to 0. issues with this solution:

    2.1. getting progress indications out of channel.read() / channel.write() is a non-trivial refactor (they are currently designed to return null or a complete object, would need to be extended)

    2.2. even with progress indication if a broker is low on memory and under a lot of load the next call to select(timeout) may return immediately (because there's sockets with data to be read/written) and we are back to busy waiting...